File: lib/apps/file_system/file_system_app.dart
import 'package:flutter/material.dart';
import '../../core/sub_app.dart';
import '../../core/search_result.dart';
import 'services/file_system_storage.dart';
import 'screens/file_browser_screen.dart';

class FileSystemApp extends SubApp {
  @override
  String get id => 'file_system';

  @override
  String get name => 'File System';

  @override
  IconData get icon => Icons.folder;

  @override
  Color get themeColor => Colors.blue;

  @override
  void onInit() {
    super.onInit();
    FileSystemStorage.instance.init();
  }

  @override
  void onDispose() {
    super.onDispose();
    FileSystemStorage.instance.close();
  }

  @override
  Widget build(BuildContext context) {
    return FileBrowserScreen();
  }

  @override
  bool get supportsSearch => true;

  @override
  Future<List<SearchResult>> search(String query) async {
    final files = await FileSystemStorage.instance.search(query);
    return files.map((file) {
      return SearchResult(
        id: file.id,
        type: SearchResultType.file,
        appId: id,
        title: file.name,
        subtitle: file.folderPath,
        preview: null,
        navigationData: {'fileId': file.id, 'folderPath': file.folderPath},
        timestamp: file.updatedAt,
      );
    }).toList();
  }

  @override
  void navigateToSearchResult(BuildContext context, SearchResult result) {
    // Navigate to the folder containing the file
    final folderPath = result.navigationData['folderPath'] as String? ?? '';
    Navigator.of(context).push(
      MaterialPageRoute(
        builder: (context) => FileBrowserScreen(initialPath: folderPath),
      ),
    );
  }
}

File: playground/Playground/Playground/Sources/Apps/FileSystem/FileSystemApp.swift
import SwiftUI
import Combine

/// File System sub-app - Document management and text extraction
class FileSystemApp: SubApp {
    let id = "fileSystem"
    let name = "Files"
    let iconName = "doc.fill"
    let themeColor = Color.orange

    let supportsSearch = true
    let supportsSharing = true
    let acceptedShareTypes = ["text/plain", "text/*", "application/json", "application/xml"]

    // ObservableObject conformance
    let objectWillChange = ObservableObjectPublisher()

    init() {}

    func buildView() -> AnyView {
        AnyView(FileListView())
    }

    // MARK: - Lifecycle

    func onInit() async {
        print("✅ File System app initialized")
    }

    // MARK: - Search

    func search(query: String) async -> [SearchResult] {
        let filesResult = FileStorage.shared.searchFiles(query: query, limit: 50)
        guard let files = filesResult.value else {
            print("❌ File search failed: \(filesResult.error?.localizedDescription ?? "unknown error")")
            return []
        }

        return files.map { file in
            let preview: String?
            if let extractedText = file.extractedText {
                let text = extractedText.replacingOccurrences(of: "\n", with: " ").trimmingCharacters(in: .whitespaces)
                preview = text.count > 100 ? String(text.prefix(100)) + "..." : text
            } else {
                preview = nil
            }

            return SearchResult(
                id: file.id,
                type: .file,
                appId: id,
                title: file.name,
                subtitle: file.formattedSize,
                preview: preview,
                navigationData: ["fileId": file.id],
                timestamp: file.createdAt
            )
        }
    }

    func navigateToSearchResult(result: SearchResult) async {
        // Navigation will be handled by the UI layer
    }

    // MARK: - Sharing

    func onReceiveShare(content: SharedContent) async {
        do {
            // Create a temporary file from shared content
            let tempDir = FileManager.default.temporaryDirectory
            let fileName = "shared_\(Date().timeIntervalSince1970).txt"
            let fileURL = tempDir.appendingPathComponent(fileName)

            try content.data.write(to: fileURL)

            // Get file info
            let fileInfo = try FileExtractionService.shared.getFileInfo(at: fileURL.path)

            // Create file in storage
            let fileResult = FileStorage.shared.createFile(
                name: fileName,
                path: fileURL.path,
                mimeType: content.type,
                sizeBytes: fileInfo.size
            )
            let file = try fileResult.get()

            // Extract text if supported
            if FileExtractionService.shared.isSupported(mimeType: content.type) {
                try await FileExtractionService.shared.extractAndUpdateFile(fileId: file.id)
            }

            print("✅ Received shared file: \(fileName)")
        } catch {
            print("❌ Failed to receive shared content: \(error)")
        }
    }
}
